### PR TITLE
feature of adding additional hdisight cluster(part)

### PR DESCRIPTION
**Pull request by voidfunction**
Not Finished. 
TBD:  decrypt the keys we get from ambri.

### CODE REVIEW

File: .idea/encodings.xml [hunk 1]
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project version="4">
+  <component name="Encoding">
+    <file url="PROJECT" charset="UTF-8" />
+  </component>
+</project>
```
Comment (guiling) [inline .idea/encodings.xml]: this file do not need be checked in

File: HdInsightSDK/src/com/microsoft/azure/hdinsight/sdk/cluster/ClusterManager.java [hunk 2]
```diff
@@ -1,120 +1,120 @@
-package com.microsoft.azure.hdinsight.sdk.cluster;
```
Comment (guiling) [inline HdInsightSDK/src/com/microsoft/azure/hdinsight/sdk/cluster/ClusterManager.java:1]: what is the difference ? what is your change

File: HdInsightSDK/src/com/microsoft/azure/hdinsight/sdk/cluster/HDInsightClusterDetail.java [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+
+import java.io.IOException;
+import java.io.NotActiveException;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Created by ltian on 12/23/2015.
+ */
+public class HDInsightClusterDetail implements IClusterDetail {
```
Comment (guiling) [inline HdInsightSDK/src/com/microsoft/azure/hdinsight/sdk/cluster/HDInsightClusterDetail.java]: try to rename it , all the clusters are HDinsight cluster, maybe ManuallyAddedClusterdetail is ok

File: HdInsightSDK/src/com/microsoft/azure/hdinsight/sdk/cluster/HDInsightClusterDetail.java [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+
+    private String clusterName;
+    private String userName;
+    private String passWord;
+
+    @Expose
```
Comment (guiling) [inline HdInsightSDK/src/com/microsoft/azure/hdinsight/sdk/cluster/HDInsightClusterDetail.java]: what is meaning of expose keyword

Comment (voidfunction) [inline HdInsightSDK/src/com/microsoft/azure/hdinsight/sdk/cluster/HDInsightClusterDetail.java]: will not add it into json string.

File: HdInsightSDK/src/com/microsoft/azure/hdinsight/sdk/storage/StorageClientImpl.java [hunk 5]
```diff
@@ -40,7 +40,8 @@ private StorageClientImpl(){}
         try {
             List<BlobContainer> bcList = new ArrayList<>();
             CloudBlobClient client = getCloudBlobClient(storageAccount);
-            for (CloudBlobContainer container : client.listContainers(null, ContainerListingDetails.ALL, null, null)) {
+            Iterable<CloudBlobContainer> temp = client.listContainers(null, ContainerListingDetails.ALL, null, null);
```
Comment (guiling) [inline HdInsightSDK/src/com/microsoft/azure/hdinsight/sdk/storage/StorageClientImpl.java]: temp is our debug name, revert this

File: src/com/microsoft/azure/hdinsight/common/CommonConst.java [hunk 6]
```diff
@@ -20,4 +20,5 @@
     public static final String AZURE_SUBSCRIPTIONS = "com.microsoft.azure.hdinsight.intellij.AzureSubscriptions";
     public static final String AZURE_USER_INFO = "com.microsoft.azure.hdinsight.intellij.AzureUserInfo";
     public static final String AZURE_USER_SUBSCRIPTIONS = "com.microsoft.azure.hdinsight.intellij.AzureUserSubscriptions";
+    public static final String HDINSIGHT_ADDITIONAL_CLUSTERS = "com.microsoft.azure.hdinsight.additionalClusters";
```
Comment (guiling) [inline src/com/microsoft/azure/hdinsight/common/CommonConst.java]: AdditionalClusters

File: src/com/microsoft/azure/hdinsight/common/HDInsightHelper.java [hunk 7]
```diff
@@ -34,6 +38,29 @@ public static HDInsightHelper getInstance(){
         return instance;
     }
 
+    private HashMap<String, ToolWindowFactory> toolWindowFactoryCollection = new HashMap<String, ToolWindowFactory>();
+
+    public void registerToolWindowFactory(String toolWindowFactoryId, ToolWindowFactory toolWindowFactory)
+    {
+        toolWindowFactoryCollection.put(toolWindowFactoryId, toolWindowFactory);
```
Comment (guiling) [inline src/com/microsoft/azure/hdinsight/common/HDInsightHelper.java]: add lock to guarantee thread-safe

File: HdInsightSDK/src/com/microsoft/azure/hdinsight/sdk/common/NotSupportExecption.java [hunk 8]
```diff
@@ -0,0 +1,7 @@
+package com.microsoft.azure.hdinsight.sdk.common;
```
Comment (guiling) [inline HdInsightSDK/src/com/microsoft/azure/hdinsight/sdk/common/NotSupportExecption.java:1]: do we really need add so many exceptions ?

File: src/com/microsoft/azure/hdinsight/serverexplore/UI/AddNewClusterFrom.java [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+    private JPanel addNewClusterPanel;
+    private JTextField clusterNameFiled;
+    private JTextField userNameField;
+    private JPasswordField passwordField;
+    private JTextField errorMessageField;
+    private JPanel buttonPanel;
+    private JButton Okbutton;
+    private JButton cancelButton;
+
+    private static final String DEFAULT_CLUSTER_ENDPOINT = "https://%s.azurehdinsight.net";
```
Comment (guiling) [inline src/com/microsoft/azure/hdinsight/serverexplore/UI/AddNewClusterFrom.java]: I remember DEFAULT_CLUSTER_ENDPOINT was referenced  more more than once, try to put it in a common string (HDInsight sdk common is a good place)

File: src/com/microsoft/azure/hdinsight/serverexplore/UI/AddNewClusterFrom.java [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+    private static final Pattern HTTPS_URL_PATTERN = Pattern.compile("https://[^/]+");
+    private static final String URL_PREFIX = "https://";
+
+    //format input string
+    private static String getClusterName(String userNameOrUrl) throws Exception {
```
Comment (guiling) [inline src/com/microsoft/azure/hdinsight/serverexplore/UI/AddNewClusterFrom.java]: 1>I think it had better to use regular express to extract cluster name
2> format this code if(){
}else{
}
3 move this to stringhelper

File: src/com/microsoft/azure/hdinsight/serverexplore/UI/AddNewClusterFrom.java [hunk 11]
```diff
@@ -0,0 +1,186 @@
...
+                            clusterName = getClusterName(clusterNameOrUrl);
+
+                            if(hdInsightRootModule.isHDInsightAdditionalClusterExist(clusterName))
+                            {
+                                errorMessage = "Cluster already exist!";
+                                isCarryOnNextStep = false;
+                            }
+
+                        } catch (Exception exception) {
+                            errorMessage = "Wrong cluster name or cluster endpoint";
```
Comment (guiling) [inline src/com/microsoft/azure/hdinsight/serverexplore/UI/AddNewClusterFrom.java]: Invalid cluster name or cluster endpoint

File: src/com/microsoft/azure/hdinsight/serverexplore/UI/AddNewClusterFrom.java [hunk 12]
```diff
@@ -0,0 +1,186 @@
...
+        this.setTitle("Add New HDInsight Cluster");
+        this.setModal(true);
+
+        addActionListener();
+    }
+
+    private void addActionListener() {
+        Okbutton.addActionListener(new ActionListener() {
+            @Override
+            public void actionPerformed(ActionEvent e) {
```
Comment (guiling) [inline src/com/microsoft/azure/hdinsight/serverexplore/UI/AddNewClusterFrom.java:64]: can you split this function ?

File: src/com/microsoft/azure/hdinsight/spark/common/StorageAccountResolveException.java [hunk 13]
```diff
@@ -0,0 +1,9 @@
+package com.microsoft.azure.hdinsight.spark.common;
```
Comment (guiling) [inline src/com/microsoft/azure/hdinsight/spark/common/StorageAccountResolveException.java:1]: remove

File: src/com/microsoft/azure/hdinsight/spark/common/UnAuthorizedException.java [hunk 14]
```diff
@@ -0,0 +1,7 @@
+package com.microsoft.azure.hdinsight.spark.common;
```
Comment (guiling) [inline src/com/microsoft/azure/hdinsight/spark/common/UnAuthorizedException.java]: remove

**Pull request merged**